### PR TITLE
Nest data

### PR DESCRIPTION
**Pull request by grantcarthew**
This PR adds support to nest the stringified objects under a root log property as discussed in #355 
It includes two tests and documentation updates.

From a performance perspective, there is an additional ternary operator, two additional concatenation operations, and a string.split(). These changes are in the `pino.js` file around line 150.

There is a new constructor option `objectKey` which defaults to `''`. Once set to a string value any objects that get logged will be nested under the key name.

This PR is a non-breaking change adding an additional option and feature.

See what you think. Happy for feedback.

### CODE REVIEW

File: pino.js [hunk 1]
```diff
@@ -152,10 +154,12 @@ function asJson (obj, msg, num) {
       if ((notHasOwnProperty || obj.hasOwnProperty(key)) && value !== undefined) {
         value = this.stringify(this.serializers[key] ? this.serializers[key](value) : value)
         if (value !== undefined) {
-          data += ',"' + key + '":' + value
+          objData += ',"' + key + '":' + value
         }
       }
     }
+    this.objectKeyString === ',"":{' || Object.keys(obj).length < 1
```
Comment (mcollina) [inline pino.js]: can you avoid calling Object.keys() for every passed object? We are already iterating over all the properties anyway, maybe we can set a boolean somewhere.

File: pino.js [hunk 2]
```diff
@@ -152,10 +154,12 @@ function asJson (obj, msg, num) {
...
+      ? data += objData : data += this.objectKeyString + objData.slice(1) + '}'
```
Comment (mcollina) [inline pino.js]: this probably reads better as a proper if/else.

File: pino.js (hunk 1 above)
Comment (jsumners) [reply pino.js]: It's really a check for `if (objData.length > 0) ...`

File: docs/API.md [hunk 3]
```diff
@@ -62,28 +62,31 @@
...
     `'trace'`; also `'silent'` is supported to disable logging. Any other value
     defines a custom level and requires supplying a level value via `levelVal`.
     Default: 'info'.
-  * `levelVal` (integer): when defining a custom log level via `level`, set to an
+  * `levelVal` (integer): When defining a custom log level via `level`, set to an
     integer value to define the new level. Default: `undefined`.
-  * `messageKey` (string): the string key for the 'message' in the JSON object. Default `msg`.
-  * `prettyPrint` (boolean|object): enables [pino.pretty](#pretty). This is intended for non-production
+  * `messageKey` (string): The string key for the 'message' in the JSON object. Default `msg`.
+  * `objectKey` (string): The root key name to nest stringified objects under. If not changed any
```
Comment (jsumners) [inline docs/API.md]: I wonder if defaulting to `undefined` would make more sense.

**Comment by davidmarkclements** [comment]
@grantcarthew could you please run the benchmarks on this branch vs master with the option set and not set

File: test/basic.test.js [hunk 4]
```diff
@@ -228,6 +228,57 @@ test('set the messageKey', function (t) {
   instance.info(message)
 })
 
+test('set the objectKey', function (t) {
```
Comment (mcollina) [inline test/basic.test.js]: This file mode has became 755. Can you restore it to 644?

File: docs/API.md [hunk 5]
```diff
@@ -38,10 +38,10 @@
 
 ### Parameters:
 + `options` (object):
-  * `safe` (boolean): avoid error caused by circular references in the object tree.
```
Comment (mcollina) [inline docs/API.md]: This file mode has become 755, can you restore as 644.

File: docs/API.md (hunk 3 above)
Comment (grantcarthew) [reply docs/API.md]: I have no preference.

**Comment by grantcarthew** [comment]
# Benchmarks

Setup is on my laptop running Win10 with Hyper-V. A Debian virtual machine is running the benchmarks.

I only ran the `basic.bench.js` and the `object.bench.js` files because they cover the two changes in output if the `objectKey` is set or unset.

## Reference `master` Branch 

```
==========
basic averages
Bunyan average: 1042.2015
Winston average: 1399.4585000000002
Bole average: 575.9385
Debug average: 726.2460000000001
LogLevel average: 654.842
Pino average: 588.0070000000001
PinoExtreme average: 298.555
==========

==========
object averages
BunyanObj average: 1195.0665
WinstonObj average: 1405.2755
BoleObj average: 624.395
LogLevelObject average: 933.739
PinoObj average: 616.3530000000001
PinoUnsafeObj average: 600.0695000000001
PinoExtremeObj average: 319.8115
PinoUnsafeExtremeObj average: 306.7065
==========
```

### Branch: `nestdata`

#### Default `objectKey` (disabled) with `Object.keys` in `if` statement

```
==========
basic averages
Bunyan average: 1116.9685
Winston average: 1349.3545
Bole average: 579.3045
Debug average: 716.7495
LogLevel average: 674.531
Pino average: 630.318
PinoExtreme average: 306.2215
==========

==========
object averages
BunyanObj average: 1090.8965
WinstonObj average: 1408.8075
BoleObj average: 632.181
LogLevelObject average: 942.5545
PinoObj average: 614.745
PinoUnsafeObj average: 606.3675000000001
PinoExtremeObj average: 318.46950000000004
PinoUnsafeExtremeObj average: 316.307
==========
```

#### With `objectKey` set to 'data' and `Object.keys` in `if` statement

```
==========
basic averages
Bunyan average: 1136.5355
Winston average: 1484.6235000000001
Bole average: 578.238
Debug average: 754.886
LogLevel average: 676.0174999999999
Pino average: 630.7764999999999
PinoExtreme average: 324.97249999999997
==========

==========
object averages
BunyanObj average: 1108.5900000000001
WinstonObj average: 1462.324
BoleObj average: 596.6134999999999
LogLevelObject average: 950.0675
PinoObj average: 646.5245
PinoUnsafeObj average: 627.765
PinoExtremeObj average: 353.0095
PinoUnsafeExtremeObj average: 366.98350000000005
==========

```

#### Default `objectKey` (disabled) with `objData.length < 1` in `if` statement

```
==========
basic averages
Bunyan average: 1105.5245
Winston average: 1441.6770000000001
Bole average: 609.4404999999999
Debug average: 793.2180000000001
LogLevel average: 685.9915
Pino average: 636.4
PinoExtreme average: 313.92650000000003
==========

==========
object averages
BunyanObj average: 1115.529
WinstonObj average: 1397.486
BoleObj average: 621.4984999999999
LogLevelObject average: 930.6735
PinoObj average: 585.8824999999999
PinoUnsafeObj average: 613.033
PinoExtremeObj average: 314.091
PinoUnsafeExtremeObj average: 310.041
==========
```

#### With `objectKey` set to 'data' and `objData.length < 1` in `if` statement

```
==========
basic averages
Bunyan average: 1050.9175
Winston average: 1329.8505
Bole average: 574.4765
Debug average: 733.5745
LogLevel average: 651.1759999999999
Pino average: 636.9200000000001
PinoExtreme average: 302.174
==========

==========
object averages
BunyanObj average: 1091.7404999999999
WinstonObj average: 1446.7965
BoleObj average: 620.108
LogLevelObject average: 960.1220000000001
PinoObj average: 633.5895
PinoUnsafeObj average: 601.7235000000001
PinoExtremeObj average: 316.5565
PinoUnsafeExtremeObj average: 328.16949999999997
==========
```

## Summary

The following summary consists of Basic being `Pino average` values and Object being `PinoObj average` values. 

| objectKey | if | Basic | Object |
|-------|------|--------|-------|
|master|master|588.007|616.353|
|not set|Object.keys|630.318|614.745|
|set|Object.keys|630.776|646.524|
|not set|objData.length < 1|636.400|585.882|
|set|objData.length < 1|636.920|633.589|

**Comment by grantcarthew** [comment]
Here is another from the master branch:
```
==========
basic averages
Bunyan average: 1044.689
Winston average: 1453.217
Bole average: 580.5905
Debug average: 775.7045
LogLevel average: 716.8885
Pino average: 630.8875
PinoExtreme average: 338.22950000000003
==========

==========
object averages
BunyanObj average: 1190.9705
WinstonObj average: 1431.1125000000002
BoleObj average: 616.0905
LogLevelObject average: 998.1120000000001
PinoObj average: 616.8109999999999
PinoUnsafeObj average: 611.0515
PinoExtremeObj average: 329.89300000000003
PinoUnsafeExtremeObj average: 344.82050000000004
==========
```

**Comment by grantcarthew** [comment]
I don't know guys. The figures vary.
Doesn't look like there is much difference between `Object.keys(objData) < 1` or `objData.length < 1`. Worth noting that the `objData` test is only carried out if the `objectKey` is set.

**Comment by davidmarkclements** [comment]
it could be that you have a lot going on in your environment, you could take 10 of each and average, or we could run it on an aws machine. Currently it looks like this change adds 40ms which I'm not sure is worth the feature

**Comment by grantcarthew** [comment]
The whole reason I started using `pino` is because `winston` was slowing done my API way too much. If this is going to affect performance then I can just wrap the function.

Thanks guys. I'll kill this for the sake of KISS.

**Pull request closed without merging**